go/ics23.go

```diff
@@ -44,11 +50,84 @@ func VerifyMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentPro
 // left and right proofs are neighbors (or left/right most if one is nil)
 // provided key is between the keys of the two proofs
 func VerifyNonMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentProof, key []byte) bool {
-	// TODO: handle batch
-	np, ok := proof.Proof.(*CommitmentProof_Nonexist)
-	if !ok {
+	// decompress it before running code (no-op if not compressed)
+	proof = Decompress(proof)
+	np := getNonExistProofForKey(proof, key)
+	if np == nil {
 		return false
 	}
-	err := np.Nonexist.Verify(spec, root, key)
+	err := np.Verify(spec, root, key)
 	return err == nil
 }
+
+// BatchVerifyMembership will ensure all items are also proven by the CommitmentProof (which should be a BatchProof,
+// unless there is one item, when a ExistenceProof may work)
+func BatchVerifyMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentProof, items map[string][]byte) bool {
+	// decompress it before running code (no-op if not compressed) - once for batch
+	proof = Decompress(proof)
+	for k, v := range items {
+		valid := VerifyMembership(spec, root, proof, []byte(k), v)
+		fmt.Printf("Validate %t\n", valid)
+		if !valid {
+			return false
+		}
+	}
+	return true
+}
+
+// BatchVerifyNonMembership will ensure all items are also proven to not be in the Commitment by the CommitmentProof
+// (which should be a BatchProof, unless there is one item, when a NonExistenceProof may work)
+func BatchVerifyNonMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentProof, keys [][]byte) bool {
```

Comment:
Should there be another function or a flag to indicate contiguous keys for which a range proof could be used (if supported by the underlying tree)?

Reply:
I would leave it as is for now (fits just the ics23 spec), but let's use this as some feedback on the spec, and then I can enhance these functions to meet the final spec.

One point is that there is no such function as `VerifyMembership(proof, key, value)`. It needs to know the context of the type of tree. You could rewrite it to example:

`VerifyMembership: (spec) => (proof, key, value) => bool`

or anything else, but capturing some spec (definition of a particular merkle format) is critical, especially for non-membership proofs.

Reply:
The idea of range/batch/neighbor calculations can definitely be expanded.

You may note that I expose non-membership proofs for stores that only themselves have a membership proof (ics23-tendermint). If the underlying store is ordered by key, we can produce all kinds of proofs, with a bunch of processing (client-side?) on the native proof format.

I'd be happy for more discussion on the ics23 spec, informed by what is built here (and what seems easy/hard to add)